core/src/main/java/com/linecorp/armeria/client/limit/DefaultConcurrencyLimit.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.limit;
+
+import static com.linecorp.armeria.client.limit.ConcurrencyLimitBuilder.noLimitFuture;
+
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.common.util.SafeCloseable;
+
+/**
+ * Concurrency settings that limits the concurrent number of active requests.
+ */
+final class DefaultConcurrencyLimit implements ConcurrencyLimit {
+
+    private final Predicate<? super ClientRequestContext> predicate;
+    private final int maxConcurrency;
+    private final int maxPendingAcquisitions;
+    private final long timeoutMillis;
+
+    private final Queue<PendingAcquisition> pendingAcquisitions = new ConcurrentLinkedQueue<>();
```

Comment:
Since there is a limit on pending acquisitions, looks like this can be `ArrayBlockingQueue`.

Reply:
Isn't `ArrayBlockingQueue` pre-allocating? I guess a user may specify a very large value for some peculiar reasons.

Reply:
>  a very large value for some peculiar reasons.

Like waiting for a reservation of Covid-19 Vaccine. 💉

Reply:
🤣